app/assets/javascripts/discourse/app/components/modal/grant-badge.js

```diff
@@ -1,80 +1,80 @@
 import { action } from "@ember/object";
-import Component from "@ember/component";
+import { empty } from "@ember/object/computed";
+import { tracked } from "@glimmer/tracking";
+import Component from "@glimmer/component";
 import Badge from "discourse/models/badge";
-import GrantBadgeController from "discourse/mixins/grant-badge-controller";
 import I18n from "I18n";
 import UserBadge from "discourse/models/user-badge";
-import discourseComputed from "discourse-common/utils/decorators";
 import { extractError } from "discourse/lib/ajax-error";
 import getURL from "discourse-common/lib/get-url";
+import {
+  grantableBadges,
+  isBadgeGrantable,
+} from "discourse/lib/grant-badge-utils";
 
-export default class GrantBadgeModal extends Component.extend(
-  GrantBadgeController
-) {
-  loading = true;
-  saving = false;
-  selectedBadgeId = null;
-  flash = null;
-  flashType = null;
-  allBadges = [];
-  userBadges = [];
+export default class GrantBadgeModal extends Component {
+  @tracked loading = true;
+  @tracked saving = false;
+  @tracked selectedBadgeId = null;
+  @tracked flash = null;
+  @tracked flashType = null;
+  @tracked allBadges = [];
+  @tracked userBadges = [];
+  @tracked availableBadges = [];
+  @empty("availableBadges") noAvailableBadges;
```

Comment:
These legacy `@ember/object/computed` decorators are not 100% compatible with modern Ember autotracking, so it's best to write this out explicitly like

```
get noAvailableBadges(){
  return !!this.availableBadges.length;
}
```